config/policies.go

```diff
@@ -0,0 +1,156 @@
+package config
+
+import (
+	"golang.ngrok.com/ngrok/internal/pb"
+)
+
+type policies struct {
+	Inbound  []*policy
+	Outbound []*policy
+}
+type policy struct {
+	Name        string
+	Expressions []string
+	Actions     []*action
+}
+type action struct {
+	Type   string
+	Config string
+}
+
+type PoliciesOption func(*policies)
+type PolicyOption func(*policy)
+type ActionOption func(*action)
+
+// WithPolicies creates a new set of policies with the provided options
+func WithPolicies(opts ...PoliciesOption) interface {
+	HTTPEndpointOption
+	TLSEndpointOption
+	TCPEndpointOption
+} {
+	p := &policies{}
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p
+}
+
+// WithInboundPolicy adds the provided policy to be applied on inbound connections on an ngrok edge.
+// The order in which policies are added is respected at runtime.
+func WithInboundPolicy(opts ...PolicyOption) PoliciesOption {
+	return func(p *policies) {
+		inP := &policy{}
+		for _, opt := range opts {
+			opt(inP)
+		}
+
+		p.Inbound = append(p.Inbound, inP)
+	}
+}
+
+// WithOutboundPolicy adds the provided policy to be applied on outbound connections on an ngrok edge.
+// The order in which policies are added is respected at runtime.
+func WithOutboundPolicy(opts ...PolicyOption) PoliciesOption {
+	return func(p *policies) {
+		outP := &policy{}
+		for _, opt := range opts {
+			opt(outP)
+		}
+		p.Outbound = append(p.Outbound, outP)
+	}
+}
+
+// WithName sets the provided name on this policy
+func WithName(name string) PolicyOption {
+	return func(p *policy) {
+		p.Name = name
+	}
+}
+
+// WithExpressions appends the provided cel expression to this policy
+func WithExpression(expr string) PolicyOption {
+	return func(p *policy) {
+		p.Expressions = append(p.Expressions, expr)
+	}
+}
+
+// WithAction appends the provided action to be executed when this policy's expressions match a connection to an ngrok edge.
+// The order in which actions are added to a policy is respected at runtime. At least one action must be specified.
+func WithAction(opts ...ActionOption) PolicyOption {
+	return func(p *policy) {
+		act := &action{}
+		for _, opt := range opts {
+			opt(act)
+		}
+		p.Actions = append(p.Actions, act)
+	}
+}
+
+// WithType sets the provided type for this action. Type must be specified.
+func WithType(typ string) ActionOption {
```

Comment:
Ditto the above comment about `PolicyOption`s. Didn't even notice at first glance that these were *Action* options lol